water/lib/Dialect/Wave/Transforms/LowerWaveToMLIR.cpp

```diff
@@ -108,24 +114,48 @@ struct LowerWaveToMLIRPass
-          mlir::populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
-              patterns, typeConverter);
-
-          wave::populateWaveMiscellaneousOpsLoweringPatterns(typeConverter,
-                                                             patterns);
-          wave::populateWaveBinaryOpLoweringPatterns(typeConverter, patterns);
-          wave::populateWaveAllocateOpLoweringPatterns(typeConverter, patterns);
-          wave::populateWaveMmaLoweringPatterns(typeConverter, patterns);
-          wave::populateWaveReadWriteLoweringPatterns(typeConverter, patterns);
-          wave::populateWaveUnaryFPOpLoweringPatterns(typeConverter, patterns);
-
-          if (failed(applyPartialConversion(op, target, std::move(patterns),
-                                            config))) {
-            op->emitError() << "failed to convert starting at this operation";
-            return WalkResult::interrupt();
+
+          // Phase 1: Convert all operations except control flow operations
+          {
+            ConversionTarget phase1Target(target);
+            // Keep iterate operations legal for now - we'll convert them in phase 2
+            phase1Target.addLegalOp<wave::IterateOp, wave::YieldOp>();
+
+            RewritePatternSet phase1Patterns(ctx);
+
+            // Add function signature conversion patterns.
+            mlir::populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
+                phase1Patterns, typeConverter);
+
+            // Add all patterns except control flow patterns
+            wave::populateWaveBinaryOpLoweringPatterns(typeConverter, phase1Patterns);
+            wave::populateWaveAllocateOpLoweringPatterns(typeConverter, phase1Patterns);
+            wave::populateWaveMmaLoweringPatterns(typeConverter, phase1Patterns);
+            wave::populateWaveReadWriteLoweringPatterns(typeConverter, phase1Patterns);
+            wave::populateWaveUnaryFPOpLoweringPatterns(typeConverter, phase1Patterns);
+
+            // Add only non-control-flow patterns from miscellaneous ops
+            wave::populateWaveNonControlFlowMiscLoweringPatterns(typeConverter, phase1Patterns);
+
+            if (failed(applyPartialConversion(op, phase1Target, std::move(phase1Patterns),
+                                              config))) {
+              op->emitError() << "failed to convert in phase 1";
+              return WalkResult::interrupt();
+            }
+          }
+
+          // Phase 2: Convert control flow operations
+          {
+            RewritePatternSet phase2Patterns(ctx);
+
+            // Add only control flow patterns
+            wave::populateWaveControlFlowLoweringPatterns(typeConverter, phase2Patterns);
+
+            if (failed(applyPartialConversion(op, target, std::move(phase2Patterns),
+                                              config))) {
+              op->emitError() << "failed to convert in phase 2";
+              return WalkResult::interrupt();
+            }
```

Comment:
I'd rather make a separate pass then, and potentially a named pipeline to manage that complexity